src/core/lib/promise/observable.h

```diff
@@ -0,0 +1,146 @@
+  void Set(T value) { state_->Set(std::move(value)); }
+
+  // Returns a promise that resolves to a T when the value becomes != current.
+  auto Next(T current) { return Observer(state_, std::move(current)); }
+
+ private:
+  // Forward declaration so we can form pointers to Observer in State.
+  class Observer;
+
+  // State keeps track of all observable state.
+  // It's a refcounted object so that promises reading the state are not tied
+  // to the lifetime of the Observable.
+  class State : public RefCounted<State> {
+   public:
+    explicit State(T value) : value_(std::move(value)) {}
+
+    // Update the value and wake all observers.
+    void Set(T value) {
+      MutexLock lock(&mu_);
+      std::swap(value_, value);
+      WakeAll();
+    }
+
+    // Export our mutex so that Observer can use it.
+    Mutex* mu() ABSL_LOCK_RETURNED(mu_) { return &mu_; }
+
+    // Fetch a ref to the current value.
+    const T& current() const ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      return value_;
+    }
+
+    // Remove an observer from the set (it no longer needs updates).
+    void Remove(Observer* observer) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      observers_.erase(observer);
+    }
+
+    // Add an observer to the set (it needs updates).
+    GRPC_MUST_USE_RESULT Waker Add(Observer* observer)
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      observers_.insert(observer);
+      return Activity::current()->MakeNonOwningWaker();
+    }
+
+   private:
+    // Wake all observers.
+    void WakeAll() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+      for (auto* observer : observers_) {
+        observer->Wakeup();
+      }
+    }
+
+    Mutex mu_;
+    // All observers that may need an update.
+    absl::flat_hash_set<Observer*> observers_ ABSL_GUARDED_BY(mu_);
+    // The current value.
+    T value_ ABSL_GUARDED_BY(mu_);
+  };
+
+  // Observer is a promise that resolves to a T when the value becomes !=
+  // current.
+  class Observer {
+   public:
+    Observer(RefCountedPtr<State> state, T current)
+        : state_(std::move(state)), current_(std::move(current)) {}
+    ~Observer() {
+      // If we saw a pending at all then we *may* be in the set of observers.
+      // If not we're definitely not and we can avoid taking the lock at all.
+      if (!saw_pending_) return;
+      MutexLock lock(state_->mu());
+      auto w = std::move(waker_);
+      state_->Remove(this);
+    }
+
+    Observer(const Observer&) = delete;
+    Observer& operator=(const Observer&) = delete;
+    Observer(Observer&& other) noexcept
+        : state_(std::move(other.state_)), current_(std::move(other.current_)) {
+      GPR_ASSERT(other.waker_.is_unwakeable());
+      GPR_ASSERT(!other.saw_pending_);
+    }
+    Observer& operator=(Observer&& other) noexcept = delete;
+
+    void Wakeup() { waker_.WakeupAsync(); }
+
+    Poll<T> operator()() {
+      MutexLock lock(state_->mu());
+      // Check if the value has changed yet.
+      if (current_ != state_->current()) {
```

Comment:
It looks like this mechanism is designed to be lossy -- i.e., if the value changes twice before this is polled again, the caller will be notified only once.  That's not necessarily a problem, since there are certainly cases for which this makes sense (e.g., resolver results).  But keep in mind that there are also other cases that need non-lossy updates (e.g., connectivity state notifications for subchannels and LB policies), so we wouldn't be able to use this mechanism for those cases.

Reply:
Right, for things that need ever value to be seen we'd use some other type, which is fine.